cleverhans/attacks.py

```diff
@@ -1,4 +1,313 @@
+from abc import ABCMeta, abstractmethod
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes. All attacks must override the
```

Comment:
It's probably best to keep comments about which methods must be overridden in the docstrings for the methods themselves. Otherwise this docstring is likely to get out of date when we change the methods and forget to also update the class docstring.